agents/go-agents/src/main/go/core/auth/auth.go

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright (c) 2012-2017 Codenvy, S.A.
+// All rights reserved. This program and the accompanying materials
+// are made available under the terms of the Eclipse Public License v1.0
+// which accompanies this distribution, and is available at
+// http://www.eclipse.org/legal/epl-v10.html
+//
+// Contributors:
+//   Codenvy, S.A. - initial API and implementation
+//
+
+// Package auth provides simple way of authentication of http requests on workspace master
+package auth
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/eclipse/che/agents/go-agents/src/main/go/core/rest"
+)
+
+// TokenCache represents authentication tokens cache
+type TokenCache interface {
+	Put(token string)           // Put adds token into the cache.
+	Expire(token string)        // Expire removes provided token from the cache.
+	Contains(token string) bool // Contains returns true if token is present in the cache and false otherwise.
+}
+
+// Handler is HTTP handler that authenticates all the http calls on workspace master.
+// Checks on workspace master if provided by request token is valid and calls ServerHTTP on delegate.
+// Otherwise if UnauthorizedHandler is configured calls ServerHTTP on it.
+// If it is not configured returns 401 with appropriate error message.
+type Handler struct {
```

Comment:
What if we change it to an internal implementation of `http.Handler` which can be created by constructing functions like
```go
func NewHandler(endpont, delegate, unauthorizedHandler) http.Handler { ... };
func NewCachingHandler(endpont, delegate, unauthorizedHandler, cache) http.Handler { ... }
```
or event handler builder.

So you can be sure that all the data required for handler is provided + you're be able to pick one of caching or non-caching handler and avoid checks like `handler.Cache != nil` in runtime.